### PR TITLE
Bind gatekeeper connect completion to the initiating browser

### PR DESCRIPTION
**Pull request by Maximo-Guk**
This PR adds a a single-use "ticket" which binds completion to the browser that started the flow. The gatekeeper's final page posts a single-use 256-bit ticket to window.opener, targeted at the Workshop's origin, so only the Workshop window that opened the popup can receive it. The Workshop activates the grant only when that ticket is redeemed.

How the ticket works:
1. Server, hashed. The Workshop keeps only the SHA-256 hash of the ticket, in the initiating user's DO (or the PendingLogin DO for sign-in), for 2m when the connect flow is completed.
2. The popup's HTML. The gatekeeper embeds the plaintext ticket in the completion page's inline script, served with Cache-Control: no-store and Referrer-Policy: no-referrer, so it isn't cached or leaked in a referrer.
3. In flight. The script calls window.opener.postMessage({type, ticket}, targetOrigin) the moment the page loads. The message is a JavaScript value handed from one window to another, it touches no storage.
4. A local variable in the opener. The Workshop's message listener reads the ticket and immediately sends it over the authenticated WebSocket RPC where it's then compared to the server's hash.

This means that if I give someone else my URL, finishing the flow gives them nothing. Their browser has no Workshop window to hand the ticket to, and even a logged-in victim's session would look it up in their own DO and find nothing. My staged grant then expires unredeemed and is revoked.

Kernel changes:
- GatekeeperConnectCallback.complete() stages the account in the user's DO under the SHA-256 of a fresh 256-bit ticket and returns the ConnectHandoff instead of persisting the account.
- AuthenticatedApi.completeConnectHandoff(ticket) redeems a ticket in the caller's own DO (single use, two-minute lifetime); an alarm revokes staged connects whose ticket never came back.
- Sign-in: LoginConnectCallbackImpl parks the minted session in PendingLogin under the ticket's hash, LoginAttempt.wait() becomes claim(ticket), so holding the attempt stub alone never yields a session. (Previously wait() handed whoever held the stub the session of whichever user finished OAuth.)

---

<a href="https://cloudflare.devinenterprise.com/review/cloudflare/cloudflare-os/pull/464" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

File: packages/mcp-shared/src/account.ts [hunk 1]
```diff
@@ -567,55 +643,104 @@ export abstract class McpAccountBase<E extends AccountEnv, P = unknown>
...
+   */
+  async commitReconnect(stageId: string): Promise<void> {
+    const staged = commitStagedCredentials<StagedReconnect>(
+      this.ctx.storage.kv, Date.now(), stageId);
+    if (!staged) throw new Error("No reconnect is awaiting confirmation. Please try again.");
+    if (staged.tokens) this.ctx.storage.kv.put<OAuthTokens>("tokens", staged.tokens);
+    // The live session was opened under the credentials being replaced, so it goes with them, as do
+    // the server record the flow observed and the registration and discovery a refresh will need.
+    this.setSessionId(staged.sessionId ?? null);
+    this.ctx.storage.kv.put<ConnectedServer>("server", staged.server);
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/mcp-shared/src/account.ts]: Fixed in 42ec03bd. The repoint branch of `beginConnect` now discards the pending stage alongside the live keys, so reconnect A's ticket fails with "No reconnect is awaiting confirmation" and the Workshop's failed-restore path leaves nothing live. As belt and braces, `commitReconnect` also refuses a stage whose own `server.endpoint` differs from the live one, so a stage that somehow outlived a move still cannot restore the old server. Test: "refuses to commit a reconnect staged before a repoint".

File: packages/mcp-shared/src/account.ts [hunk 2]
```diff
@@ -279,12 +318,22 @@ export abstract class McpAccountBase<E extends AccountEnv, P = unknown>
 
   async prepareReconnect(initiationNonce: string): Promise<void> {
     this.advanceConnectionGeneration();
-    this.ctx.storage.kv.put("reconnecting", true);
     this.ctx.storage.kv.put("expiredNotified", false);
+    this.ctx.storage.kv.delete(RECONNECT_TOKENS_KEY);
+    // The flow works on copies of the client registration and discovery, so what it learns (or the
+    // SDK invalidates) stays off the live keys until the commit.
+    for (const [live, scratch] of [
+      ["oauthClient", RECONNECT_CLIENT_KEY], ["oauthDiscovery", RECONNECT_DISCOVERY_KEY],
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/mcp-shared/src/account.ts]: Fixed in 42ec03bd. `prepareReconnect` no longer copies the live discovery into the reconnect scratch key, only the client registration; discovery runs again from the probe's current challenge, and the provider's `matchesIssuer` drops the copied client when the discovered issuer differs, so a moved authorization server gets a fresh registration too. Note this predates the PR: the live `oauthDiscovery` was reused on reconnect on main as well. Test: "rediscovers the authorization server on reconnect".

File: packages/workshop-frontend/src/components/auth/OAuthButtons.tsx [hunk 3]
```diff
@@ -33,61 +47,103 @@ export default function OAuthButtons({ rpcStub, vendors, onSuccess }: OAuthButto
...
-        pollRef.current = window.setInterval(() => {
-          if (popup.closed) finish(() => reject(new Error('Sign-in was cancelled.')))
+        // Claims may overlap: a foreign ticket answered with null must not hold up the real one
+        // behind it, and `finish` settles only once.
+        function claimTicket(ticket: string) {
+          if (settled) return
+          stopPolling()
+          attempt.claim(ticket)
+            .then(t => { if (t !== null) finish(() => resolve(t)) })
+            .catch(e => finish(() => reject(e instanceof Error ? e : new Error('Could not sign in'))))
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-frontend/src/components/auth/OAuthButtons.tsx:140]: Fixed in da7e8584. The poll is hoisted into `startPolling()` and a claim that resolves `null` restarts it, so closing the popup after a foreign ticket hands the buttons back as before. The existing "keeps waiting when a broadcast ticket belongs to another attempt" test now closes the popup after the foreign claim and asserts the button re-enables.

File: packages/workshop-frontend/src/connectHandoff.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      const ticket = connectHandoffTicket(event)
+      if (ticket !== null) redeem(ticket, event.source as Window | null)
+    }
+    window.addEventListener('message', onMessage)
+    const channel = 'BroadcastChannel' in globalThis
+      ? new BroadcastChannel(CONNECT_HANDOFF_MESSAGE_TYPE)
+      : null
+    channel?.addEventListener('message', (event: MessageEvent) => {
+      const ticket = parseHandoffEnvelope(event.data)
+      if (ticket !== null) redeem(ticket, null)
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-frontend/src/connectHandoff.ts]: Fixed in da7e8584, with a per-tab `sessionStorage` marker: `openConnectWindow` sets it and the listener redeems a broadcast ticket only in the tab that holds it, clearing it on the first redemption. It is per-tab and reload-stable, so the initiating tab still completes after a reload while its siblings stay silent; the trade-off is that a connect whose tab was closed expires like an abandoned one. The `message` transport is unchanged since its source is the popup this tab holds, and a tab without storage falls back to redeeming everything. This predates today's commits (it dates from the round-2 BroadcastChannel fallback).

**Review by kentonv — APPROVED** [review]

File: packages/workshop-backend/src/auth/login-flow.ts [hunk 5]
```diff
@@ -8,74 +8,117 @@
...
+  /**
+   * Release the token to the holder of the matching ticket. A ticket that is not this attempt's
+   * (the window may hear every same-origin broadcast) yields null and leaves the result in place.
+   * Single use otherwise: the ticket is hashed before the read, so the read, check and removal of a
+   * matching result happen in one step under the input gate and a repeat gets no second try.
+   */
+  async claim(ticket: string): Promise<string | null> {
+    const hash = /^[0-9a-f]{64}$/.test(ticket) ? await hashSecret(Uint8Array.fromHex(ticket)) : null;
+    const result = this.ctx.storage.kv.get<PendingResult>(RESULT_KEY);
+    if (!result || Date.now() >= result.expiresAt) {
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-backend/src/auth/login-flow.ts:120]: Fixed in cfe74391. `startGatekeeperLogin` now writes a `{pending: true}` marker to the PendingLogin DO before the gatekeeper flow starts, with the connect-nonce lifetime (10 minutes, `LOGIN_PENDING_LIFETIME_MS`) rather than the 2-minute result lifetime

File: packages/mcp-shared/src/account.ts [hunk 6]
```diff
@@ -567,55 +654,109 @@ export abstract class McpAccountBase<E extends AccountEnv, P = unknown>
...
+    if (!this.isCurrentConnection(server, pending.generation)) return null;
+    const tokens = this.freshTokens(reconnect);
     if (!tokens) throw new Error("The authorization server returned no access token.");
     this.ctx.storage.kv.delete("oauthVerifier");
 
-    const info = await this.probe(server, tokens.access_token, pending.generation);
-    if (!this.isCurrentConnection(server, pending.generation)) return false;
-    await this.complete(server, info, pending.generation);
-    return true;
+    const probed = await this.probe(server, tokens.access_token);
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/mcp-shared/src/account.ts]: Fixed in 577bdd1f. The probe and `complete()` now run under a try/catch in `acceptAuthCode`

File: packages/gatekeeper-cloudflare/src/cloudflare.ts [hunk 7]
```diff
@@ -319,7 +329,20 @@ export class UserAccount extends DurableObject<Env> {
...
     }
-    return true;
+    return handoff;
+  }
+
+  /** Makes the grant staged under `stageId` live; see GatekeeperUser.commitReconnect. */
+  async commitReconnect(stageId: string): Promise<void> {
+    const grant = commitStagedCredentials<StoredGrant>(this.ctx.storage.kv, Date.now(), stageId);
+    if (!grant) throw new Error("No reconnect is awaiting confirmation. Please try again.");
+    this.#writeGrant(grant);
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/gatekeeper-cloudflare/src/cloudflare.ts:339]: This is pre-existing, deferring

File: packages/workshop-backend/src/auth/login-flow.ts [hunk 8]
```diff
@@ -124,11 +198,13 @@ export class LoginConnectCallbackImpl
...
       if (this.ctx.props.vendorId === CLOUDFLARE_VENDOR_ID) {
-        await userStub.linkConnectedAccountFromLogin(account, this.ctx.props.vendorId, expiresAt);
+        const accountId = await userStub.linkConnectedAccountFromLogin(
+            account, this.ctx.props.vendorId, expiresAt);
+        await pending.link(userStub.id.toString(), accountId);
       }
       // Session tokens are "<doName>:<secret>"; PublicApi.authenticate() routes via idFromName of
       // the first part. The user DO is keyed by email, so the prefix must be the email.
-      await pending.deliver(`${email}:${secret}`);
+      await pending.deliver(`${email}:${secret}`, ticketHash);
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-backend/src/auth/login-flow.ts:207]: This is pre-existing and it doesn't matter much it's just an orphaned hash if someone abandons their sign-in, and we currently record succesful signin's anyways and never expire them. If we actually care about storage we should add some pruning here eventually

File: packages/workshop-frontend/src/connectHandoff.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    window.addEventListener('message', onMessage)
+    const channel = 'BroadcastChannel' in globalThis
+      ? new BroadcastChannel(CONNECT_HANDOFF_MESSAGE_TYPE)
+      : null
+    channel?.addEventListener('message', (event: MessageEvent) => {
+      const ticket = parseHandoffEnvelope(event.data)
+      if (ticket === null || !hasPendingConnect()) return
+      // One connect, one redemption: the marker is spent whether or not the server accepts.
+      clearPendingConnect()
+      redeem(ticket, null)
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-frontend/src/connectHandoff.ts:179]: Fixed in 577bdd1f marker now records when the popup was opened and is spent only by a successful  redemption

**Comment by Maximo-Guk** [comment]
> **Findings**
> 
> 1. **P1** `packages/workshop-backend/src/user.ts:1780`: Expiry calls full provider revocation. GitHub and Google revoke the entire app grant, so a phished authorization can revoke a victim’s existing legitimate connection. Unconfirmed cleanup needs a local-only credential discard.
> 2. **P2** `packages/workshop-backend/src/user.ts:1823`: One pending record with an unavailable gatekeeper stub aborts the whole sweep. Later abandoned grants remain indefinitely. Store list-safe expiry metadata separately and load payloads independently.
> 3. **P2** `packages/workshop-backend/src/auth/login-flow.ts:184`: Cloudflare callbacks retain only an account ID. Re-sign-in replaces the capability under that ID, so a delayed expiry callback from the old account can incorrectly mark the new grant expired. Bind callbacks to an account generation.
> 
> [github run](/cloudflare/cloudflare-os/actions/runs/34381461826)

1. Fixed in 577bdd1f, for GitHub. DELETE /applications/{id}/grant revokes every token the user holds for the app, so dropping an abandoned pending connect took the same user's working connection down. The  gatekeeper now calls DELETE /applications/{id}/token
2. Deferred. The sweep is retried every 2 minutes, so a poisoned record just delays later ones
3. Deferred. Account-id reuse under a Cloudflare re-sign-in predates this PR

File: packages/workshop-shared/src/gatekeeper.ts [hunk 10]
```diff
@@ -534,7 +584,21 @@ export interface GatekeeperConnectCallback extends WorkerEntrypoint {
    * operation to fail. If not provided, the system relies on the gatekeeper calling
    * `credentialsExpired()` when a refresh or authorization failure is detected.
    */
-  complete(user: Fetcher<GatekeeperUser>, expiresAt?: Date): Promise<void>;
+  complete(user: Fetcher<GatekeeperUser>, expiresAt?: Date): Promise<ConnectHandoff>;
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-shared/src/gatekeeper.ts:587]: This is ok

File: packages/workshop-frontend/src/connectHandoff.ts [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    const redeem = (ticket: string, source: Window | null) => {
+      authenticatedApi.completeConnectHandoff(ticket).then(
+        () => {
+          if (source) {
+            source.close?.()
+            return
+          }
+          clearPendingConnect()
+          // oxlint-disable-next-line unicorn/require-post-message-target-origin -- a BroadcastChannel has no targetOrigin.
+          channel?.postMessage({ type: CONNECT_HANDOFF_ACK_MESSAGE_TYPE, ticket })
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-frontend/src/connectHandoff.ts:162]: I'm going to defer this

File: packages/mcp-shared/src/account.ts [hunk 12]
```diff
@@ -279,12 +320,24 @@ export abstract class McpAccountBase<E extends AccountEnv, P = unknown>
 
   async prepareReconnect(initiationNonce: string): Promise<void> {
     this.advanceConnectionGeneration();
-    this.ctx.storage.kv.put("reconnecting", true);
     this.ctx.storage.kv.put("expiredNotified", false);
+    this.ctx.storage.kv.delete(RECONNECT_TOKENS_KEY);
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/mcp-shared/src/account.ts:328]: This is pre-existing, will defer for now, the orphaned grant is held by nobody anyways

File: packages/mcp-shared/src/account.ts [hunk 13]
```diff
@@ -567,55 +654,161 @@ export abstract class McpAccountBase<E extends AccountEnv, P = unknown>
...
+    }
+    if (staged.tokens) this.ctx.storage.kv.put<OAuthTokens>("tokens", staged.tokens);
+    // The live session was opened under the credentials being replaced, so it goes with them, as do
+    // the server record the flow observed and the registration and discovery a refresh will need.
+    this.setSessionId(staged.sessionId ?? null);
+    this.ctx.storage.kv.put<ConnectedServer>("server", staged.server);
+    if (staged.client) this.ctx.storage.kv.put("oauthClient", staged.client);
+    else this.ctx.storage.kv.delete("oauthClient");
+    if (staged.discovery) this.ctx.storage.kv.put("oauthDiscovery", staged.discovery);
+    else this.ctx.storage.kv.delete("oauthDiscovery");
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/mcp-shared/src/account.ts]: Fixed in 566eee2b, commitReconnect now revokes the live tokens with the discovery and client they were issued under (the shared revokeTokens path), then drops them, before the commit deletes that discovery

File: packages/workshop-frontend/src/components/auth/OAuthButtons.tsx [hunk 14]
```diff
@@ -33,61 +49,121 @@ export default function OAuthButtons({ rpcStub, vendors, onSuccess }: OAuthButto
...
-          .then(t => finish(() => resolve(t)))
-          .catch(e => finish(() => reject(e instanceof Error ? e : new Error('Could not sign in'))))
+        // Claims may overlap: a foreign ticket answered with null must not hold up the real one
+        // behind it, and `finish` settles only once. Polling pauses during a claim so a popup that
+        // closes itself on completion is not read as a cancellation, and resumes after a foreign
+        // ticket, or closing the popup afterwards would leave the buttons stuck.
+        function claimTicket(ticket: string) {
+          if (settled) return
+          stopPolling()
+          attempt.claim(ticket)
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/workshop-frontend/src/components/auth/OAuthButtons.tsx:128]: Going to defer this

**Comment by Maximo-Guk** [comment]
> **Findings**
> 
> 1. **P2** `packages/workshop-frontend/src/connectHandoff.ts:104`: The ownership marker expires before a valid OAuth handoff necessarily completes. OAuth receives a fresh 10-minute nonce, and callback processing can continue after validation, so a valid ticket arriving just after the marker’s 10-minute cutoff is ignored until the staged grant expires.
> 2. **P2** `packages/mcp-shared/src/account.ts:675`: An OAuth reconnect reconstructs the server from live storage, losing updated portal metadata supplied before redirect. Renaming a portal while retaining its endpoint results in a successful reconnect that restores the old name. Persist the flow’s `ConnectedServer` in `PendingAuthorization`.
> 
> Tests were unavailable because `pnpm` is not installed in the runner.
> 
> [github run](/cloudflare/cloudflare-os/actions/runs/34398651554)

1. Fixed in 8092b32b. The marker now lives 30 minutes
2. Portal rename also fixed in 8092b32b, The record the flow resolved before the redirect now travels inPendingAuthorization.server, and acceptAuthCode stages from it rather than rebuilding from the live record

**Comment by Maximo-Guk** [comment]
> **Findings**
> 
> 1. **P2** `packages/workshop-frontend/src/connectHandoff.ts:101`: If `sessionStorage.setItem()` hits quota while reads still work, no marker exists and valid handoffs are ignored. Keep an in-memory fallback marker.
> 2. **P2** `packages/workshop-frontend/src/routes/__root.tsx:114`: Authenticated `/signup` renders outside the handoff listener. Navigating there during a connection causes the ticket to expire. Keep the listener mounted whenever authenticated.
> 
> Posted 1 inline finding covering staged credentials not being revoked on disconnect.
> 
> Tests not run because `pnpm` is unavailable.

1. This is a really big edge case, and not really worth fixing now. I'm sure other things will fail if they hit their browser storage quota.
2. This doesn't really matter, if a logged in user navigates to signup during connect flow they can always retry later and is an edge case

File: packages/gatekeeper-github/src/github.ts [hunk 15]
```diff
@@ -1304,34 +1308,47 @@ export class UserAccount extends DurableObject<Env> {
...
-    if (reconnecting) {
-      this.ctx.storage.kv.delete("reconnecting");
-      await callback.credentialsRestored();
+    let handoff: ConnectHandoff;
+    if (stored.reconnect) {
+      // The reconnect URL is a bearer capability, so the new grant is only staged until the Workshop
+      // has confirmed the browser that finished the flow is the owner's (see commitReconnect). Bound
+      // gadgets keep reading the current token meanwhile. The stage id ties the Workshop's ticket to
+      // this grant, so an overlapping reconnect cannot be committed by it.
+      const stageId = stageCredentials(this.ctx.storage.kv, grant, Date.now());
```
[the remark this answers is not in the document]
Comment (Maximo-Guk) [reply packages/gatekeeper-github/src/github.ts:1317]: This is similar to https://github.com/cloudflare/cloudflare-os/pull/464#discussion_r3972488483 and doesn't really matter that much we can address in the future
